shared/util/tf/tf_parser.go

```diff
@@ -0,0 +1,348 @@
+					return errors.New("Terraform code has errors")
+				}
+
+				variables = append(variables, *v)
+			}
+		}
+		return err
+	})
+	if err != nil {
+		return variables, errors.Wrap(err, "failed to parse terraform files")
+	}
+	return variables, nil
+}
+
+// A slimmed down borrowing of https://github.com/hashicorp/terraform/blob/b81253999e83eedc70da79bae2455e15c6d44b74/internal/configs/named_values.go#L51
+func decodeVariableBlock(block *hcl.Block) (*Variable, hcl.Diagnostics) {
+	v := &Variable{
+		Name:      block.Labels[0],
+		DeclRange: block.DefRange,
+	}
+
+	content, diags := block.Body.Content(variableBlockSchema)
+
+	if attr, exists := content.Attributes["type"]; exists {
+		ty, tyDefaults, tyDiags := decodeVariableType(attr.Expr)
+		diags = append(diags, tyDiags...)
+		v.ConstraintType = ty
+		v.TypeDefaults = tyDefaults
+		v.Type = ty.WithoutOptionalAttributesDeep()
+	}
+
+	if attr, exists := content.Attributes["default"]; exists {
+		val, valDiags := attr.Expr.Value(nil)
+		diags = append(diags, valDiags...)
+
+		if v.ConstraintType != cty.NilType {
+			var err error
+			if v.TypeDefaults != nil && !val.IsNull() {
+				val = v.TypeDefaults.Apply(val)
+			}
+			val, err = convert.Convert(val, v.ConstraintType)
+			if err != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid default value for variable",
+					Detail:   fmt.Sprintf("This default value is not compatible with the variable's type constraint: %s.", err),
+					Subject:  attr.Expr.Range().Ptr(),
+				})
+				val = cty.DynamicVal
+			}
+		}
+
+		v.Default = val
+	}
+
+	return v, diags
+}
+
+func decodeVariableType(expr hcl.Expression) (cty.Type, *typeexpr.Defaults, hcl.Diagnostics) {
+	switch hcl.ExprAsKeyword(expr) {
+	case "list":
+		return cty.List(cty.DynamicPseudoType), nil, nil
+	case "map":
+		return cty.Map(cty.DynamicPseudoType), nil, nil
+	default:
+	}
+
+	ty, typeDefaults, diags := typeexpr.TypeConstraintWithDefaults(expr)
+	if diags.HasErrors() {
+		return cty.DynamicPseudoType, nil, diags
+	}
+
+	return ty, typeDefaults, diags
+}
+
+func (tf TfParser) ParseOutputs(dir string) ([]Output, error) {
+	outputs := []Output{}
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			if info.Name() == ".terraform" || info.Name() == ".git" {
+				return filepath.SkipDir
+			}
+		}
+		if !info.IsDir() && filepath.Ext(path) == ".tf" {
+			b, err := os.ReadFile(path)
+			if err != nil {
+				return err
+			}
+
+			f, diags := hclsyntax.ParseConfig(b, path, hcl.Pos{Line: 1, Column: 1})
+			if diags.HasErrors() {
+				return errors.Wrapf(diags.Errs()[0], "failed to parse %s", path)
+			}
+
+			content, _, diags := f.Body.PartialContent(outputBlockSchema)
+			if diags.HasErrors() {
+				return errors.New("Terraform code has errors")
+			}
+			for _, block := range content.Blocks {
+				if block.Type != "output" {
+					continue
```

Comment:
I feel like there is a lot repeated code between these methods

Reply:
I will save refactoring for a separate PR. Flow here seems similar but it isn't. I most likely will combine all of the ParseX() methods into one, but this requires a lot of testing.